File: app/src/main/java/knf/kuma/favorite/FavsSectionAdapter.java
package knf.kuma.favorite;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.animeinfo.ActivityAnime;
import knf.kuma.commons.PatternUtil;
import knf.kuma.commons.PicassoSingle;
import knf.kuma.commons.PrefsUtil;
import knf.kuma.favorite.objects.InfoContainer;
import knf.kuma.pojos.FavoriteObject;

public class FavsSectionAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    static final int TYPE_HEADER = 0;
    static final int TYPE_ITEM = 1;

    private Context context;
    private Fragment fragment;
    private RecyclerView recyclerView;
    private OnMoveListener listener;
    private List<FavoriteObject> list = new ArrayList<>();

    public FavsSectionAdapter(Fragment fragment, RecyclerView recyclerView) {
        this.fragment = fragment;
        this.listener = (OnMoveListener) fragment;
        this.context = fragment.getContext();
        this.recyclerView = recyclerView;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            default:
            case TYPE_HEADER:
                return new HeaderHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_fav_header, parent, false));
            case TYPE_ITEM:
                return new ItemHolder(LayoutInflater.from(parent.getContext()).inflate(getLayout(), parent, false));
        }
    }

    @LayoutRes
    private int getLayout() {
        if (PrefsUtil.getLayType().equals("0")) {
            return R.layout.item_fav;
        } else {
            return R.layout.item_fav_grid;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder h, int position) {
        FavoriteObject object = list.get(position);
        if (h instanceof HeaderHolder) {
            HeaderHolder holder = (HeaderHolder) h;
            holder.header.setText(object.name);
            holder.action.setOnClickListener(v -> listener.onEdit(object.name));
        } else if (h instanceof ItemHolder) {
            ItemHolder holder = (ItemHolder) h;
            PicassoSingle.get(context).load(PatternUtil.getCover(object.aid)).into(holder.imageView);
            holder.title.setText(object.name);
            holder.type.setText(object.type);
            holder.cardView.setOnClickListener(view -> ActivityAnime.open(fragment, object, holder.imageView));
            holder.cardView.setOnLongClickListener(v -> {
                listener.onSelect(object);
                return true;
            });
        }
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public int getItemViewType(int position) {
        return list.get(position).isSection ? TYPE_HEADER : TYPE_ITEM;
    }

    public void updatePosition(InfoContainer container) {
        list = container.updated;
        recyclerView.post(() -> notifyItemMoved(container.from, container.to));
    }

    public void updateList(List<FavoriteObject> list) {
        this.list = list;
        recyclerView.post(this::notifyDataSetChanged);
    }

    interface OnMoveListener {
        void onSelect(FavoriteObject object);

        void onEdit(String category);
    }

    class ItemHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.card)
        CardView cardView;
        @BindView(R.id.img)
        ImageView imageView;
        @BindView(R.id.title)
        TextView title;
        @BindView(R.id.type)
        TextView type;

        public ItemHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    class HeaderHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.header)
        TextView header;
        @BindView(R.id.action)
        ImageButton action;

        public HeaderHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/knf/kuma/favorite/FavsSectionAdapter.kt
package knf.kuma.favorite

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.RecyclerView
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView
import knf.kuma.R
import knf.kuma.ads.AdCallback
import knf.kuma.ads.AdCardItemHolder
import knf.kuma.ads.AdsUtilsMob
import knf.kuma.animeinfo.ActivityAnime
import knf.kuma.commons.DesignUtils
import knf.kuma.commons.PrefsUtil
import knf.kuma.commons.bind
import knf.kuma.commons.load
import knf.kuma.pojos.av1.FavoriteAV1
import knf.kuma.pojos.av1.FavoriteBase
import knf.kuma.pojos.av1.FavoriteSectionAV1
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


class FavsSectionAdapter(private val fragment: Fragment, private val recyclerView: FastScrollRecyclerView, private val showSections: Boolean) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), FastScrollRecyclerView.SectionedAdapter {

    private val context: Context?
    private val listener: OnMoveListener
    private val orderType = PrefsUtil.favsOrder
    var list: List<FavoriteBase> = ArrayList()

    private val layout: Int
        @LayoutRes
        get() = if (PrefsUtil.layType == "0") {
            if (DesignUtils.isFlat) R.layout.item_fav_material else R.layout.item_fav
        } else {
            if (DesignUtils.isFlat) R.layout.item_fav_grid_material else R.layout.item_fav_grid
        }

    init {
        this.listener = fragment as OnMoveListener
        this.context = fragment.context
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_ITEM -> ItemHolder(LayoutInflater.from(parent.context).inflate(layout, parent, false))
            TYPE_AD -> AdCardItemHolder(parent, AdCardItemHolder.TYPE_FAV).also {
                it.loadAd(fragment.lifecycleScope, object : AdCallback {
                    override fun getID(): String = AdsUtilsMob.FAVORITE_BANNER
                }, 500)
            }
            else -> HeaderHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_fav_header, parent, false))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val favoriteObject = list[position]
        if (holder is HeaderHolder && favoriteObject is FavoriteSectionAV1) {
            holder.header.text = favoriteObject.name.let {
                if (it == FavoriteAV1.CATEGORY_NONE) "Sin categoria" else it
            }
            if (favoriteObject.name == FavoriteAV1.CATEGORY_NONE) {
                holder.action.isVisible = false
            } else {
                holder.action.isVisible = true
                holder.action.setOnClickListener { listener.onEdit(favoriteObject.name) }
            }
        } else if (holder is ItemHolder && favoriteObject is FavoriteAV1) {
            holder.imageView.load(favoriteObject.imageUrl())
            holder.title.text = favoriteObject.name
            holder.type.text = favoriteObject.typeText()
            holder.cardView.setOnClickListener { ActivityAnime.open(fragment, favoriteObject, holder.imageView) }
            if (showSections)
                holder.cardView.setOnLongClickListener {
                    listener.onSelect(favoriteObject)
                    true
                }
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun getItemViewType(position: Int): Int {
        return try {
            if (list[position] is FavoriteSectionAV1) TYPE_HEADER else TYPE_ITEM
        } catch (_: Exception) {
            TYPE_ITEM
        }
    }

    override fun getSectionName(position: Int): String {
        return try {
            if (showSections)
                ""
            else {
                val item = list[position] as FavoriteAV1
                when (orderType) {
                    0 -> {
                        val name = item.name
                        if (name.isNotEmpty())
                            name.substring(0, 1).uppercase()
                        else
                            name
                    }

                    else -> item.aid.toString()
                }
            }
        } catch (_: IllegalStateException) {
            ""
        }
    }

    fun updateList(list: List<FavoriteBase>) {
        fragment.lifecycleScope.launch(Dispatchers.IO) {
            this@FavsSectionAdapter.list = list
            recyclerView.post { this@FavsSectionAdapter.notifyDataSetChanged() }
        }
    }

    internal interface OnMoveListener {
        fun onSelect(favoriteObject: FavoriteAV1)

        fun onEdit(category: String)
    }

    internal class ItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val cardView: View by itemView.bind(R.id.card)
        val imageView: ImageView by itemView.bind(R.id.img)
        val title: TextView by itemView.bind(R.id.title)
        val type: TextView by itemView.bind(R.id.type)
    }

    internal class HeaderHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val header: TextView by itemView.bind(R.id.header)
        val action: ImageButton by itemView.bind(R.id.action)
    }

    companion object {
        internal const val TYPE_HEADER = 0
        internal const val TYPE_ITEM = 1
        internal const val TYPE_AD = 2
    }
}
